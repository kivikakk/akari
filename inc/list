// This file is part of Akari.
// Copyright 2010 Arlen Cuss
//
// Akari is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Akari is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Akari.  If not, see <http://www.gnu.org/licenses/>.

#ifndef __LIST
#define __LIST

namespace std {
	template <typename T> class list {
	protected:
		struct listNode;
	public:
		explicit list(): _begin(0), _end(0)
		{ }

		list(const list &r): _begin(0), _end(0) {
			for (typename list::iterator it = r.begin(); it != r.end(); ++it)
				push_back(*it);
		}

		~list() {
			clear();
		}

		class base_iterator {
		public:
			bool operator ==(const base_iterator &r) const {
				return _data == r._data;
			}

			bool operator !=(const base_iterator &r) const {
				return _data != r._data;
			}

			T &operator *() {
				 return *_data->item;
			}

			const T &operator *() const {
				 return *_data->item;
			}

			T *&operator ->() {
				return _data->item;
			}

			T *const &operator ->() const {
				return _data->item;
			}

		protected:
			base_iterator(listNode *data) {
				_data = data;
			}

			listNode *_data;
		};

		class iterator : public base_iterator {
		friend class list;
		public:
			iterator &operator ++() {
				this->_data = this->_data->next;
				return *this;
			}

			iterator &operator --() {
				this->_data = this->_data->prev;
				return *this;
			}

		protected:
			iterator(listNode *data): base_iterator(data)
			{ }
		};

		class reverse_iterator : public base_iterator {
		friend class list;
		public:
			reverse_iterator &operator ++() {
				this->_data = this->_data->prev;
				return *this;
			}

			reverse_iterator &operator --() {
				this->_data = this->_data->next;
				return *this;
			}

		protected:
			reverse_iterator(listNode *data): base_iterator(data)
			{ }
		};

		iterator begin() {
			return iterator(_begin);
		}

		const iterator begin() const {
			return iterator(_begin);
		}

		iterator end() {
			return iterator(0);
		}

		const iterator end() const {
			return iterator(0);
		}

		reverse_iterator rbegin() {
			return reverse_iterator(_end);
		}

		const reverse_iterator rbegin() const {
			return reverse_iterator(_end);
		}

		reverse_iterator rend() {
			return reverse_iterator(0);
		}

		const reverse_iterator rend() const {
			return reverse_iterator(0);
		}

		u32 size() const {
			u32 len = 0;
			for (iterator it = begin(); it != end(); ++it, ++len);
			return len;
		}

		/*
		const T &operator[](u32 index) const {
			iterator t = begin();
			while (index--)
				++t;
			return *t;
		}

		T &operator[](u32 index) {
			iterator t = begin();
			while (index--)
				++t;
			return *t;
		}
		*/

		T &push_back(const T &data) {
			listNode **writeHead = &_begin, *lastNode = 0;
			while (*writeHead) {
				lastNode = *writeHead;
				writeHead = &(*writeHead)->next;
			}

			listNode *newNode = new listNode;
			newNode->item = new T(data);
			newNode->next = 0;
			newNode->prev = lastNode;

			*writeHead = newNode;
			_end = newNode;

			return *newNode->item;
		}

		void pop_front() {
			listNode *old = _begin;
			_begin = _begin->next;
			_begin->prev = 0;
			delete old->item;
			delete old;
		}

		void remove(const T &data) {
			listNode **writeHead = &_begin, *lastNode = 0;
			while (*writeHead) {
				if (*(*writeHead)->item == data) {
					listNode *old = *writeHead;
					*writeHead = (*writeHead)->next;
					if (*writeHead)
						(*writeHead)->prev = lastNode;
					if (_end == old) _end = lastNode;
					delete old->item;
					delete old;
					break;
				}
				lastNode = *writeHead;
				writeHead = &(*writeHead)->next;
			}
		}

		void clear() {
			listNode *traverse = _begin;
			while (traverse) {
				listNode *next = traverse->next;
				delete traverse->item;
				delete traverse;
				traverse = next;
			}
			_begin = 0;
			_end = 0;
		}

		bool empty() const {
			return !_begin;
		}

	protected:
		struct listNode {
			T *item;
			listNode *next, *prev;
		};

		listNode *_begin, *_end;
	};
}

#endif

// vim: filetype=cpp:
