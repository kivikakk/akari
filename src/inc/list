// This file is part of Akari.
// Copyright 2010 Arlen Cuss
//
// Akari is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.
//
// Akari is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with Akari.  If not, see <http://www.gnu.org/licenses/>.

#ifndef __LIST
#define __LIST

namespace std {
	template <typename T> class list {
	protected:
		struct listNode;
	public:
		explicit list(): _begin(0), _end(0)
		{ }

		list(const list &r): _begin(0), _end(0) {
			for (typename list::iterator it = r.begin(); it != r.end(); ++it)
				push_back(*it);
		}

		list &operator =(const list &r) {
			clear();
			for (iterator it = r.begin(); it != r.end(); ++it)
				push_back(*it);

			return *this;
		}

		~list() {
			clear();
		}

		class base_iterator {
		friend class list;
		public:
			base_iterator(const base_iterator &r): _data(r._data)
			{ }

			virtual ~base_iterator() { }

			base_iterator &operator =(const base_iterator &r) {
				_data = r._data;
				return *this;
			}

			bool operator ==(const base_iterator &r) const {
				return _data == r._data;
			}

			bool operator !=(const base_iterator &r) const {
				return _data != r._data;
			}

			T &operator *() {
				 return *_data->item;
			}

			const T &operator *() const {
				 return *_data->item;
			}

			T *&operator ->() {
				return _data->item;
			}

			T *const &operator ->() const {
				return _data->item;
			}

		protected:
			base_iterator(listNode *data): _data(data)
			{ }

			listNode *_data;
		};

		class iterator : public base_iterator {
		friend class list;
		public:
			iterator &operator ++() {
				this->_data = this->_data->next;
				return *this;
			}

			iterator &operator --() {
				// XXX: I'm not sure if this is even tenable.
				// _end doesn't exist, which is the whole problem.
				// We don't have a pointer to our actual list. Hence
				// if we're the end() iterator, _data is 0 and we
				// basically are hanging out in the middle of nowhere.
				// Do we want one?
				if (this->_data)
					this->_data = this->_data->prev;
				else
					this->_data = _end;
				return *this;
			}

		protected:
			iterator(listNode *data): base_iterator(data)
			{ }
		};

		class reverse_iterator : public base_iterator {
		friend class list;
		public:
			reverse_iterator &operator ++() {
				this->_data = this->_data->prev;
				return *this;
			}

			reverse_iterator &operator --() {
				if (this->_data)
					this->_data = this->_data->next;
				else
					this->_data = _begin;
				return *this;
			}

		protected:
			reverse_iterator(listNode *data): base_iterator(data)
			{ }
		};

		iterator begin() {
			return iterator(_begin);
		}

		const iterator begin() const {
			return iterator(_begin);
		}

		iterator end() {
			return iterator(0);
		}

		const iterator end() const {
			return iterator(0);
		}

		reverse_iterator rbegin() {
			return reverse_iterator(_end);
		}

		const reverse_iterator rbegin() const {
			return reverse_iterator(_end);
		}

		reverse_iterator rend() {
			return reverse_iterator(0);
		}

		const reverse_iterator rend() const {
			return reverse_iterator(0);
		}

		u32 size() const {
			u32 len = 0;
			for (iterator it = begin(); it != end(); ++it, ++len);
			return len;
		}

		/*
		const T &operator[](u32 index) const {
			iterator t = begin();
			while (index--)
				++t;
			return *t;
		}

		T &operator[](u32 index) {
			iterator t = begin();
			while (index--)
				++t;
			return *t;
		}
		*/

		iterator insert(const base_iterator &pos, const T &data) {
			// amortised constant time
			if (pos._data == 0) {
				// End of list
				push_back(data);
				return iterator(_end);
			}

			if (pos._data->prev == 0) {
				// Start of list
				push_front(data);
				return begin();
			}

			listNode *newNode = new listNode;
			newNode->item = new T(data);
			newNode->next = pos._data;
			newNode->prev = pos._data->prev;
			newNode->next->prev = newNode;
			newNode->prev->next = newNode;

			return iterator(newNode);
		}

		iterator erase(const base_iterator &pos) {
			// amortised constant time (given we're erasing one)
			listNode *old = pos._data;
			if (old->prev == 0) {
				pop_front();
				return begin();
			}

			if (old->next == 0) {
				pop_back();
				return end();
			}

			iterator ret = iterator(old->next);
			old->prev->next = old->next;
			old->next->prev = old->prev;
			delete old->item;
			delete old;

			return ret;
		}

		void push_front(const T &data) {
			// amortised constant time
			listNode *newNode = new listNode;
			newNode->item = new T(data);
			newNode->next = _begin;
			newNode->prev = 0;

			if (newNode->next)
				newNode->next->prev = newNode;
			else
				_end = newNode;

			_begin = newNode;
		}

		void push_back(const T &data) {
			// amortised constant time
			listNode *newNode = new listNode;
			newNode->item = new T(data);
			newNode->next = 0;
			newNode->prev = _end;

			if (newNode->prev)
				newNode->prev->next = newNode;
			else
				_begin = newNode;

			_end = newNode;
		}

		void pop_front() {
			// amortised constant time
			listNode *old = _begin;
			_begin = _begin->next;

			if (_begin)
				_begin->prev = 0;
			else
				_end = 0;

			delete old->item;
			delete old;
		}

		void pop_back() {
			// amortised constant time
			listNode *old = _end;
			_end = _end->prev;

			if (_end)
				_end->next = 0;
			else
				_begin = 0;

			delete old->item;
			delete old;
		}

		void remove(const T &data) {
			// linear in size of list
			// XXX: this should remove any and all "data"s found;
			// currently only removes one.
			listNode **writeHead = &_begin, *lastNode = 0;
			while (*writeHead) {
				if (*(*writeHead)->item == data) {
					listNode *old = *writeHead;
					*writeHead = (*writeHead)->next;
					if (*writeHead)
						(*writeHead)->prev = lastNode;
					if (_end == old) _end = lastNode;
					delete old->item;
					delete old;
					break;
				}
				lastNode = *writeHead;
				writeHead = &(*writeHead)->next;
			}
		}

		void clear() {
			listNode *traverse = _begin;
			while (traverse) {
				listNode *next = traverse->next;
				delete traverse->item;
				delete traverse;
				traverse = next;
			}
			_begin = 0;
			_end = 0;
		}

		bool empty() const {
			return !_begin;
		}

	protected:
		struct listNode {
			T *item;
			listNode *next, *prev;
		};

		listNode *_begin, *_end;
	};
}

#endif

// vim: filetype=cpp:
