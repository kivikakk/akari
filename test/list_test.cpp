#include <cstdio>

typedef unsigned long u32;

#include "list"

typedef std::list<int> list_type;
list_type my_list;

void report() {
	printf("my_list(%u): ", (unsigned int)my_list.size());
	for (list_type::iterator it = my_list.begin(); it != my_list.end(); ++it) {
		if (it != my_list.begin())
			printf(", ");
		printf("%d", *it);
	}
	printf("\n");
}

int main() {
	my_list.push_back(4);
	my_list.push_back(3);
	my_list.push_back(2);
	my_list.push_back(1);
	report();

	my_list.insert(my_list.begin(), 1);
	report();

	my_list.insert(++++my_list.begin(), 2);
	report();

	my_list.insert(++++++++my_list.begin(), 3);
	report();

	my_list.insert(++++++++++++my_list.begin(), 4);
	report();

	my_list.insert(my_list.end(), 0);
	report();

	my_list.insert(my_list.begin(), 0);
	report();
}
