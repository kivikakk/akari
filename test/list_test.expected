my_list(4): 4, 3, 2, 1
my_list(5): 1, 4, 3, 2, 1
my_list(6): 1, 4, 2, 3, 2, 1
my_list(7): 1, 4, 2, 3, 3, 2, 1
my_list(8): 1, 4, 2, 3, 3, 2, 4, 1
my_list(9): 1, 4, 2, 3, 3, 2, 4, 1, 0
my_list(10): 0, 1, 4, 2, 3, 3, 2, 4, 1, 0

